discretisedfield/field.py

```diff
@@ -3479,3 +3480,210 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+                                  name=name,
+                                  attrs=dict(units=units))
+
+        for dim in geo_units_dict:
+            data_array[dim].attrs['units'] = geo_units_dict[dim]
+
+        return data_array
+
+    @classmethod
+    def from_xarray(cls, xa):
+        if not isinstance(xa, xr.DataArray):
+            msg = "Argument must be a xr.DataArray."
+            raise TypeError(msg)
+
+        if xa.ndim not in [3, 4]:
+            msg = "DataArray dimensions must be 3 or 4."
+            raise ValueError(msg)
+
+        if xa.ndim == 3:
+            for dim in xa.dims:
+                if dim not in ['x', 'y', 'z']:
+                    msg = "The dimensions must be either 'x', 'y', or 'z'."
+                    raise ValueError(msg)
+        elif xa.ndim == 4:
+            for dim in xa.dims:
+                if dim not in ['x', 'y', 'z', 'comp']:
+                    msg = "The dimensions must be either "\
+                          "'x', 'y', 'z' or 'comp'."
+                    raise ValueError(msg)
+            if xa['comp'].values.size != len(set(xa['comp'].values)):
+                msg = "All the components should be unique."
+                raise ValueError(msg)
+            if 'comp' in xa.coords:
+                for comp in xa['comp'].values:
+                    if not isinstance(comp, str):
+                        msg = "The values of 'comp' must be a string."
+                        raise ValueError(msg)
+                    elif comp not in ['x', 'y', 'z']:
+                        msg = "The values of 'comp' must be either "\
+                              "'x', 'y', or 'z'."
+                        raise ValueError(msg)
+
+        for i in 'xyz':
+            if xa[i].values.size != 1:
+                if not np.allclose(np.diff(xa[i].values),
+                                   np.diff(xa[i].values).mean()):
+                    msg = f'Co-ordinates of {i} must be'\
+                          ' equally spaced.'
+                    raise ValueError(msg)
+
+        if xa.values.dtype not in [np.float64, np.complex128]:
+            msg = "DataArray values must be either "\
+                  "np.float64 or np.complex128."
+            raise ValueError(msg)
+
+        cell = tuple()
+        p1 = tuple()
+        p2 = tuple()
+        for i in 'xyz':
+            dec = 5  # minimum decimal places to round to.
+
+            if xa[i].values.size > 1:
+                diff = np.diff(xa[i].values).mean()
+            else:
+                diff = np.float64(1e-9)  # imput value for cell dimension.
+
+            exp = np.log10(np.abs(diff))
+            if exp < 0:
+                dec = dec + int(np.abs(exp).round())
+
+            diff = diff.round(decimals=dec)
+            cell = cell + (abs(diff),)
+            p1 = p1 + ((xa[i].values[0] - diff * 0.5).round(decimals=dec),)
+            p2 = p2 + ((xa[i].values[-1] + diff * 0.5).round(decimals=dec),)
```

Comment:
Do we have to round?
```python
try:
    cell = np.array([xa[i].values[1] - xa[i].values[0] for i in 'xyz'])
except IndexError:
    cell = xa.attrs['cell']  # maybe wrong syntax (?)
p1 = np.array([xa[i].values[0] for i in 'xyz']) - cell / 2
p2 = np.array([xa[i].values[-1] for i in 'xyz']) + cell / 2
```

Reply:
I initially did not round. But the values of `p1`, `p2`, and `cell` of original mesh and the mesh re-created by `from_xarray`do not match. I observed deviations like `0` becoming `1.02753e-23` and `5e-9` becoming `5.0000000000001e-9` after performing the floating point arithmetic.

These deviations give error when comparing the original field and the one re-created by `from_xarray` (even with `field.allclose()` method) as the undying mesh do not match. This error is mitigated by rounding but then the issue is to what decimal place to round.

Reply:
Could we use `dec = np.finfo(xa.values.dtype).precision`?

Reply:
Yes, this works perfectly! I have changed it.